Allow maintaining the blockfilterindex when using prune

jonasschnelli opened the pull request:
Maintaining the blockfilterindexes in prune mode is possible and may lead to efficient p2p based rescans of wallets (restore backups, import/sweep keys) beyond the prune height (rescans not part of that PR).

This PR allows running the blockfilterindex(es) in conjunction with pruning.
* Bitcoind/Qt will shutdown during startup when missing block data has been detected ([re]enable `-blockfilterindex` when we already have pruned)
* manual block pruning is disabled during blockfilterindex sync
* auto-pruning is delayed during blockfilterindex sync

ToDos:
* [x] Functional tests

jonasschnelli committed:
Allow blockfilter in conjunction with prune

promag commented on src/index/base.cpp:
```diff
@@ -66,6 +66,29 @@ bool BaseIndex::Init()
+            // it looks like this index is not built yet
+            // make sure we have all data back to the genesis
+            const CBlockIndex* block = chainActive.Tip();
+            while (block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA)) {
```
0d8f7369cd8b51dfb7b3b9b920e4e11f7962faa9 

This doesn't check `status & BLOCK_HAVE_DATA` on the tip, is that intentional?

MarcoFalke replied:
I don't think the tip can ever not have data?

jonasschnelli replied:
> I don't think the tip can ever not have data? 

Jup.

promag commented on src/index/base.cpp:
```diff
@@ -70,21 +70,21 @@ bool BaseIndex::Init()
-            // it looks like this index is not built yet
-            // make sure we have all data back to the genesis
+            // index is not built yet
+            // make sure we have all block data back to the genesis
```
d2ab62fede34b7980863617f3669c019ee42d029 

These comment were added in the previous commit, could fixup these hunks?

promag commented on src/index/base.h:
```diff
@@ -100,6 +100,8 @@ class BaseIndex : public CValidationInterface
     /// not block and immediately returns false.
     bool BlockUntilSyncedToCurrentChain();
 
+    bool IsSynced();
```
8bb1ca8a276f5f6fc5af3639c2d58c496b71bf8f 

`bool IsSyned() const;`. Any reason to not write implementation here?

jonasschnelli replied:
Fixed.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1004,6 +1004,13 @@ static UniValue pruneblockchain(const JSONRPCRequest& request)
     LOCK(cs_main);
 
+    // reject manual pruning in case if a blockfilterindex is syncing
+    ForEachBlockFilterIndex([](BlockFilterIndex& index) {
```
bd11606a95022478e6ef1ccaada1589cff13a327

Can be checked before locking `cs_main`?

jonasschnelli replied:
Fixed.

promag commented on src/rpc/blockchain.cpp:
```diff
@@ -1004,6 +1004,13 @@ static UniValue pruneblockchain(const JSONRPCRequest& request)
+    // reject manual pruning in case if a blockfilterindex is syncing
+    ForEachBlockFilterIndex([](BlockFilterIndex& index) {
+        if (!index.IsSynced()) {
+            throw JSONRPCError(RPC_MISC_ERROR, "Cannot prune blocks because blockfilterindex is syncing.");
```
bd11606a95022478e6ef1ccaada1589cff13a327

Do you see a way to test this?

jonasschnelli replied:
I'm currently working on making the block-/undo-filesize dynamic which would then allow testing this more reliable.

luke-jr replied:
Would be nicer to just check if the indexes are beyond the requested prune height...

promag commented on src/validation.cpp:
```diff
@@ -2085,23 +2086,34 @@ bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &
+    // we delay pruning when one of the blockfilterindex is syncing
+    bool delay_pruning = false;
+    ForEachBlockFilterIndex([&delay_pruning](BlockFilterIndex& index) {
+        if (!index.IsSynced()) { delay_pruning = true; }
```
bf744288bd58718844841c088c498cd57a8f87d8 

nit, no need for `{}`. Alternative:
```
delay_pruning |= !index.IsSynced();
```

jonasschnelli replied:
Rather not do `|= !` for code readability, but removed the extra unnecessary block

promag commented on test/lint/lint-circular-dependencies.sh:
```diff
@@ -11,6 +11,7 @@ export LC_ALL=C
 EXPECTED_CIRCULAR_DEPENDENCIES=(
     "chainparamsbase -> util/system -> chainparamsbase"
     "index/txindex -> validation -> index/txindex"
+    "index/blockfilterindex -> validation -> index/blockfilterindex"
```
bf744288bd58718844841c088c498cd57a8f87d8

An alternative is to add a global `g_delay_pruning` in validation.h which gets updated every time an index `m_synced` is updated.

MarcoFalke replied:
Or move the "load block" function into a separate "storage" module.

jonasschnelli replied:
The global `g_delay_pruning` smells fragile. Better not do that.

I like the "storage" module approach which though would be something for another PR.

promag commented:
Closes #15867?

jonasschnelli commented:
Rebased and fixed reported points.

luke-jr commented on src/index/base.cpp:
```diff
@@ -23,7 +23,7 @@ static void FatalError(const char* fmt, const Args&... args)
     SetMiscWarning(strMessage);
     LogPrintf("*** %s\n", strMessage);
     uiInterface.ThreadSafeMessageBox(
-        "Error: A fatal internal error occurred, see debug.log for details",
```
Aren't we intentionally keeping the message vague here to avoid confusing users?

luke-jr commented on src/index/base.cpp:
```diff
@@ -66,6 +66,28 @@ bool BaseIndex::Init()
+        }
+        // in case the index has a best block set and is not fully synced
+        // check if we have the required blocks to continue building the index
+        else if (!(m_best_block_index.load()->nStatus & BLOCK_HAVE_DATA)) {
```
IIRC, having the data for block N is not a guarantee we have it for block N+M, since blocks are stored out of order and pruned only as entire files.

luke-jr replied:
If `m_best_block_index` is not in the main chain, we may need its parent blocks to undo indexing. This should be checked here too.

luke-jr commented on src/index/base.cpp:
```diff
@@ -178,6 +200,10 @@ bool BaseIndex::Rewind(const CBlockIndex* current_tip, const CBlockIndex* new_ti
+    // Pruning has a minimum of 288 blocks-to-keep and getting the index
+    // out of sync may be possible but a users fault.
+    // In case we reorg beyond the pruned depth, ReadBlockFromDisk would
+    // throw and led to a graceful shutdown
```
lead*

luke-jr commented on src/validation.cpp:
```diff
@@ -2105,23 +2106,34 @@ bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &
+    bool delay_pruning = false;
+    ForEachBlockFilterIndex([&delay_pruning](BlockFilterIndex& index) {
+        if (!index.IsSynced()) delay_pruning = true;
+    });
```
Won't this effectively prevent all pruning until IBD completes? Seems important to prune as soon as all indexes are caught up to the given height...

Roasbeef commented:
Any chance this will be revived? This is a blocker for `lnd` to support a pruned `bitcoind` node, as we want to be able to fetch filters from `bitcoind`, then manually fetch blocks ourselves (if `bitcoind` doesn't have them) to scan them for rescans or just normal wallet sync.

prusnak commented:
This feature increases the value of running a pruned node significantly. +1 for adding this!

luke-jr commented:
I suspect #19463 could be useful for this

jonasschnelli committed:
Add debug startup parameter -fastprune for more effective pruning tests

jonasschnelli committed:
Add functional test for blockfilterindex in prune-mode

jonasschnelli committed:
Avoid pruning below the blockfilterindex sync height

jonasschnelli committed:
Avoid accessing nullpointer in BaseIndex::GetSummary()

jonasschnelli committed:
Add "index/blockfilterindex -> validation -> index/blockfilterindex" to expected circular dependencies

jonasschnelli commented:
Rewrote this PR and tried to fix the reported point by @promag, @MarcoFalke and @luke-jr.
Added `-smallblockfiles` debug parameter to test pruning more effectively (added a functional-test as well).

fjahr reviewed (commented):
Concept ACK although I am still thinking about how to keep the chances low that users shoot themselves in the foot and unnecessarily need to redownload the blockchain again.

It isn't great that the changes in `FlushStateToDisk()` aren't tested. I am thinking about how to make that possible.

jonasschnelli commented on src/validation.cpp:
```diff
@@ -2236,17 +2237,30 @@ bool CChainState::FlushStateToDisk(
         bool fFlushForPrune = false;
         bool fDoFullFlush = false;
+
+        // make sure we don't prune below the blockfilterindexes bestblocks
```
Good point. Will change.

jonasschnelli commented on src/validation.cpp:
```diff
@@ -2236,17 +2237,30 @@ bool CChainState::FlushStateToDisk(
                 LOG_TIME_MILLIS_WITH_CATEGORY("find files to prune (manual)", BCLog::BENCH);
 
-                m_blockman.FindFilesToPruneManual(setFilesToPrune, nManualPruneHeight, m_chain.Height());
+                m_blockman.FindFilesToPruneManual(setFilesToPrune, has_blockfilter ? std::max(nManualPruneHeight, index_lowest_blockheight) : nManualPruneHeight, m_chain.Height());
```
Absolutely. Fixed.

jonasschnelli commented on src/validation.cpp:
```diff
@@ -2236,17 +2237,30 @@ bool CChainState::FlushStateToDisk(
+            if (height < index_lowest_blockheight) {
+                height = index_lowest_blockheight;
+            }
+            has_blockfilter = false;
```
Thanks. Fixed.

jonasschnelli commented on src/validation.cpp:
```diff
@@ -2236,17 +2237,30 @@ bool CChainState::FlushStateToDisk(
+        ForEachBlockFilterIndex([&index_lowest_blockheight, &has_blockfilter](BlockFilterIndex& index) {
+            int height = index.GetSummary().best_block_height;
+            if (height < index_lowest_blockheight) {
+                height = index_lowest_blockheight;
```
Thanks. Fixed.

jonasschnelli commented:
Thanks @fjahr for the reviews.
And I agree that there should be a test for the FlushStateToDisk() changes.

achow101 commented:
Code Review ACK 0280274368e2df7c7832549cd581c710875a6be1

This is simpler than I expected it to be.

prusnak commented on src/validation.cpp:
```diff
@@ -3203,7 +3217,7 @@ static bool FindBlockPos(FlatFilePos &pos, unsigned int nAddSize, unsigned int n
     bool finalize_undo = false;
     if (!fKnown) {
-        while (vinfoBlockFile[nFile].nSize + nAddSize >= MAX_BLOCKFILE_SIZE) {
+        while (vinfoBlockFile[nFile].nSize + nAddSize >= (gArgs.GetBoolArg("-fastprune", false) ? 0x10000 /* 64kb */ : MAX_BLOCKFILE_SIZE)) {
```
The numeric constant could be extracted into `src/validation.h` (same file where `MAX_BLOCKFILE_SIZE` is defined).

jonasschnelli replied:
I thought about it but since its a debug only parameter I rather keep that inline.

prusnak commented on src/validation.cpp:
```diff
@@ -3986,7 +4000,7 @@ void BlockManager::FindFilesToPrune(std::set<int>& setFilesToPrune, uint64_t nPr
 static FlatFileSeq BlockFileSeq()
 {
-    return FlatFileSeq(GetBlocksDir(), "blk", BLOCKFILE_CHUNK_SIZE);
+    return FlatFileSeq(GetBlocksDir(), "blk", gArgs.GetBoolArg("-fastprune", false) ? 0x4000 /* 16kb */ : BLOCKFILE_CHUNK_SIZE);
```
The numeric constant could be extracted into `src/validation.h` (same file where `BLOCKFILE_CHUNK_SIZE ` is defined).

benthecarman reviewed (approved):
ACK 0280274368e2df7c7832549cd581c710875a6be1

ryanofsky commented on src/validation.cpp:
```diff
@@ -2246,17 +2247,30 @@ bool CChainState::FlushStateToDisk(
                 LOG_TIME_MILLIS_WITH_CATEGORY("find files to prune", BCLog::BENCH);
 
-                m_blockman.FindFilesToPrune(setFilesToPrune, chainparams.PruneAfterHeight(), m_chain.Height(), IsInitialBlockDownload());
+                m_blockman.FindFilesToPrune(setFilesToPrune, has_blockfilter ? std::min(chainparams.PruneAfterHeight(), (uint64_t)index_lowest_blockheight) : chainparams.PruneAfterHeight(), m_chain.Height(), IsInitialBlockDownload());
```
In commit "Avoid pruning below the blockfilterindex sync height" (bda1c9fc90a704d1962b16fe4242b5637c930b29)

I don't think just passing a lower `nPruneAfterHeight` argument value is sufficient to avoid pruning here. I think the `index_lowest_blockheight` variable (`last_prune` variable in my previous suggestion) has to be passed as a new argument to `FindFilesToPrune`. If you look at the `FindFilesToPrune` implementation you can see it is not using `nPruneAfterHeight` value at all to compute `nLastBlockWeCanPrune`, which is the value we need to change there.

It's also easy to see this code is wrong because it will have no effect at all if index_lowest_blockheight is bigger than chainparams.PruneAfterHeight, which is 100,000.

ryanofsky reviewed (commented):
Looks close, but I think there are some bugs, see comments

- [ ] 92b4a9cc3eca0897e455e00144a96e88b5c4ea1c Allow blockfilter in conjunction with prune (1/6)
- [X] a779d94b709bbc063025ae3aa1b47b5a5f67eea0 Avoid accessing nullpointer in BaseIndex::GetSummary() (2/6)
- [ ] bda1c9fc90a704d1962b16fe4242b5637c930b29 Avoid pruning below the blockfilterindex sync height (3/6)
- [X] e601442d541ad9e42496f511b328fae755a15555 Add debug startup parameter -fastprune for more effective pruning tests (4/6)
- [X] 87ba09d0d980b6d3468df3661ecff02030255db7 Add functional test for blockfilterindex in prune-mode (5/6)
- [X] 0280274368e2df7c7832549cd581c710875a6be1 Add "index/blockfilterindex -> validation -> index/blockfilterindex" to expected circular dependencies (6/6)

jonasschnelli commented on src/validation.cpp:
```diff
@@ -2246,17 +2247,30 @@ bool CChainState::FlushStateToDisk(
         if (fPruneMode && (fCheckForPruning || nManualPruneHeight > 0) && !fReindex) {
+            // make sure we don't prune below the blockfilterindexes bestblocks
+            // pruning is height-based
+            int index_lowest_blockheight = m_chain.Height();
+            bool has_blockfilter = false;
+            ForEachBlockFilterIndex([&index_lowest_blockheight, &has_blockfilter](BlockFilterIndex& index) {
+                int height = index.GetSummary().best_block_height;
+                if (height < index_lowest_blockheight) {
+                    index_lowest_blockheight = height;
+                }
+                has_blockfilter = true;
+            });
```
Using `last_prune` instead of `m_chain.Height()` will lead to preserve 288 blocks below the artificial tip (`last_prune`) which IMO are not subject to reorg.

ryanofsky replied:
> Using `last_prune` instead of `m_chain.Height()` will lead to preserve 288 blocks below the artificial tip (`last_prune`) which IMO are not subject to reorg.

If you're talking about the non-manual case, the you need to FindFilesToPrune a new parameter, not use the existing one, see https://github.com/bitcoin/bitcoin/pull/15946#discussion_r573391858. Not suggesting preserving below the tip.

jonasschnelli replied:
Yes. You're right. Lets just accept the 288blocks buffer for now (can be changed later if necessary).

jonasschnelli commented:
Thanks a lot for testing @ryanofsky.
Addressed your points (+rebased)(force pushed).

ryanofsky commented on src/validation.cpp:
```diff
@@ -2249,17 +2250,25 @@ bool CChainState::FlushStateToDisk(
                 LOG_TIME_MILLIS_WITH_CATEGORY("find files to prune", BCLog::BENCH);
 
-                m_blockman.FindFilesToPrune(setFilesToPrune, chainparams.PruneAfterHeight(), m_chain.Height(), IsInitialBlockDownload());
+                m_blockman.FindFilesToPrune(setFilesToPrune, chainparams.PruneAfterHeight(), last_prune, IsInitialBlockDownload());
```
In commit "Avoid pruning below the blockfilterindex sync height" (14c95089fe9727794c1a6101d5cedaca54292bb7)

Pretending `last_prune` as the chain height may be safe, because the only side effect is to not prune blocks that could be pruned, but it is confusing, and also creates a difference in behavior between the manual prune case which uses the real height and this case which pretends the min index sync height is the height.

If this behavior is going to be kept, there should be an explanatory comment because it looks broken, and not obviously broken in a safe way instead of a dangerous way.

But better than adding a comment would be to make the code straightforward: Stop passing `last_prune` as `chain_tip_height`. Instead, continue passing `m_chain.Height()` as `chain_tip_height`, pass `last_prune` as `prune_height`, and assign `nLastBlockWeCanPrune = min(prune_height, chain_tip_height - MIN_BLOCKS_TO_KEEP)` inside `FindFilesToPrune`. This would be more consistent with the logic in `FindFilesToPruneManual`, as well as more efficient, and more obviously correct.

jonasschnelli replied:
Yes. That makes sense. Implemented now as suggested by @ryanofsky.

ryanofsky reviewed (approved):
Code review ACK e2251dfb05f79a7b52aff51773759d1929fd9ade. Only change is rebase and implementing some suggestions.

Please consider implementing the suggestion below. It is a small 2 line change which I think would make the non-manual prune code less confusing, more efficient, and more consistent with the manual prune code.

fjahr commented on test/functional/feature_blockfilterindex_prune.py:
```diff
@@ -0,0 +1,45 @@
+        self.log.info("make sure we get an init error when starting the node again with block filters")
+        self.nodes[1].assert_start_raises_init_error(extra_args=self.extra_args[1])
+        self.nodes[1].assert_debug_log(["basic block filter index best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)"])
+
```
Suggested addition

```suggestion
        self.log.info("make sure the node starts again with the -reindex arg")
        reindex_args = self.extra_args[1]
        reindex_args.append("-reindex")
        self.start_node(1, extra_args=reindex_args)
```

jonasschnelli replied:
Added that additional test.

fjahr reviewed (commented):
Code review ACK e2251df

I agree with @ryanofsky that his suggestion would be an improvement, at least a comment makes sense and the update to the test description noted below would be good. If you decide to make these changes I will do my best to re-review quickly.

jonasschnelli commented on test/functional/feature_blockfilterindex_prune.py:
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test RPC misc output."""
```
Thanks for spotting. Fixed.

jonasschnelli commented on test/functional/feature_blockfilterindex_prune.py:
```diff
@@ -0,0 +1,45 @@
+
+    def run_test(self):
+        # test basic pruning compatibility & filter access of pruned blocks
+        self.log.info("check if we can access a blockfilter when pruning is enabled but no actual pruned blocks")
```
Fixed.

jonasschnelli commented:
Thanks @ryanofsky and @fjahr! Fixed the reported points. Thanks for a quick retest.

ryanofsky reviewed (approved):
Code review ACK 84716b134e9afca2fba7731de4af1f22fa1b1518. Only changes since last review were suggested new FindFilesToPrune argument and test.

Change is neat and straightforward and I think close to ready for merge. 

This should also get release notes at some point since it adds a new feature and a new error condition that wasn't checked before.

benthecarman reviewed (approved):
tACK 84716b134e9afca2fba7731de4af1f22fa1b1518

fjahr commented:
Code review ACK 84716b1

Checked that only changes since last review were small improvements on the functional test and the added parameter to `FindFilesToPrune()`.

jnewbery commented on src/validation.cpp:
```diff
@@ -17,6 +17,7 @@
 #include <cuckoocache.h>
 #include <flatfile.h>
 #include <hash.h>
+#include <index/blockfilterindex.h>
```
Is this making blockfilterindex part of validation code?

fanquake commented:
This change was part of 22.0, removing "Needs release note".
